grove/measurements/estimation.py

```diff
@@ -0,0 +1,161 @@
+"""
+Utilities for estimating expected values of Pauli terms given pyquil programs
+"""
+import numpy as np
+from pyquil.paulis import PauliSum, PauliTerm, commuting_sets, sI
+from pyquil.quil import Program
+from pyquil.gates import RY, RX
+
+
+class CommutationError(Exception):
+    pass
+
+
+def remove_imaginary_terms(pauli_sums):
+    """
+    Remove the imaginary components of the Pauli sum term
+
+    :param pauli_sums:
+    :return:
+    """
+    if not isinstance(pauli_sums, PauliSum):
+        raise TypeError("not a pauli sum. please give me one")
+    new_term = sI(0) * 0.0
+    for term in pauli_sums:
+        if (not np.isclose(term.coefficient.real, 0.0) and
+                np.isclose(term.coefficient.imag, 0.0)):
+            new_term += term
+
+    return new_term
+
+
+def get_rotation_program(pauli_term):
+    """
+    Generate a rotation program so that the pauli term is diagoanl
+
+    :param pauli_term:
+    :return: rotaiton program
+    :rtype: Program
+    """
+    meas_basis_change = Program()
+    for index, gate in pauli_term:
+        if gate == 'X':
+            meas_basis_change.inst(RY(-np.pi / 2, index))
+        elif gate == 'Y':
+            meas_basis_change.inst(RX(np.pi / 2, index))
+        elif gate == 'Z':
+            pass
+        else:
+            raise ValueError()
+
+    return meas_basis_change
+
+
+def get_parity(pauli_terms, bitstring_results):
+    """
+    Calculate the parity at each of the marked qubits in pauli_terms
+
+    Each list of bits in the bitstring_results is the maximum weight pauli term
+    in pauli_terms.  The bits in each element of bitstring_results correspond
+    to a numerical ordering of the maximal weight element of pauli_terms.
+
+    An example:
+
+    A maximum weight pauli term is Z1Z5Z6.
+
+    each element of bitstring_results is something in
+    :math:`\{0, 1\}^{\otimes 3}` where the first bit corresponds to the measured
+    value of qubit 1, the second bit corresponds to the measured value of qubit
+    5, and the third bit corresponds to the measured value of qubit 6.
+
+    :param pauli_terms:
+    :param bitstring_results:
+    :return:
+    """
```

Comment:
This docstring has a few typos and could generally use some love. Would it be possible to expand the example with a bitstring and the corresponding parity? Is the parity simply the eigenvalue of a PauliTerm (consisting only of Z's and I's) for a given bitstring?

Reply:
Yeah, I'm including the clarifying example text: 
```
"""
   Consider a Pauli term Z1 Z5 Z6 I7 and a collection of single-qubit
    measurement results corresponding to measurements in the z-basis on qubits
    {1, 5, 6, 7}. Each element of bitstring_results is an element of
    :math:`\{0, 1\}^{\otimes 4}`.  If [0, 0, 1, 0] and [1, 0, 1, 1]
    are the two projective measurement results in `bitstring_results` then
    this method returns a 1 x 2 numpy array with values [[-1, 1]]
"""
```